src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -918,15 +922,22 @@ void GcsActorManager::ReconstructActor(
   }
 }
 
-void GcsActorManager::OnActorCreationFailed(std::shared_ptr<GcsActor> actor,
-                                            bool destroy_actor) {
-  if (destroy_actor) {
-    DestroyActor(actor->GetActorID());
-  } else {
+void GcsActorManager::OnActorSchedulingFailed(std::shared_ptr<GcsActor> actor,
+                                              bool runtime_env_setup_failed) {
+  if (!runtime_env_setup_failed) {
     // We will attempt to schedule this actor once an eligible node is
     // registered.
     pending_actors_.emplace_back(std::move(actor));
+    return;
   }
+
+  auto death_cause = std::make_shared<rpc::ActorDeathCause>();
+  // TODO(sang, lixin) 1. Make this message more friendly 2. Show this message in
+  // object.get()'s error.
+  death_cause->mutable_runtime_env_setup_failure_context()->set_error_message(
+      "RuntimeEnv setup failed!");
+  // If there is runtime env failure, mark this actor as dead immediately.
+  ReconstructActor(actor->GetActorID(), /*need_reschedule=*/false, death_cause);
```

Comment:
But some cleanup is already done when the actor is created. We don't need to clean it again. ReconstructActor is good enough here.

Reply:
Isn’t DestroyActor already idempotent? 

I got your point, I think when reading code, this can easily cause confusion (like why not DestroyActor). When you described the high level function, it sounds easier to understand to use DestroyActor because we don’t need to understand if actor data has cleaned up or not at this point

Reply:
OK I agree with you. DestroyActor is easier to understand here. I'll change it.